### PR TITLE
triangle time peepoHappy

### PR DESCRIPTION
**Pull request by AiMikoto**

### CODE REVIEW

File: src/graphics/constants.h [hunk 1]
```diff
@@ -1,5 +1,12 @@
 #pragma once
+#include <vector>
+#include <string>
+
 using uint = unsigned int;
 namespace graphics {
-	
+	//Too bad we're not using c++ 20, or we could just use vectors and strings in constexpr,
```
Comment (iarba) [inline src/graphics/constants.h]: :disappointed:

File: src/graphics/gl_functions.cpp [hunk 2]
```diff
@@ -54,16 +58,149 @@ GLFWwindow * graphics::createGLFWContext(int width, int height, std::string name
...
+std::string gph::ShaderLoader::readShaderFile(std::string path) {
+	std::string res;
+	std::ifstream fileStream(path, std::ios::in);
+	if (fileStream.is_open()) {
+		std::string line = "";
+		while (std::getline(fileStream, line))
+			res += line + "\n";
+		fileStream.close();
+	}
+	else throw "Impossible to open " + path;
```
Comment (iarba) [inline src/graphics/gl_functions.cpp]: try throwing exceptions instead of strings

Comment (AiMikoto) [reply src/graphics/gl_functions.cpp]: yeah I wanted to, wanted to check it works properly first, will change it

File: CMakeLists.txt [hunk 3]
```diff
@@ -58,6 +63,7 @@ add_executable(instance ${INSTANCE_SOURCES})
 add_executable(gfx ${GRAPHICS_SOURCES})
 target_link_libraries(gfx glfw)
 target_link_libraries(gfx glm)
+target_link_libraries(gfx Assimp)
```
Comment (iarba) [inline CMakeLists.txt]: `target_link_libraries(gfx assimp)`
as in - no capital letter required - that breaks compilation

File: src/graphics/shader.h [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+		Shader();
+		~Shader();
+		void draw();
+		void setUniform(const GLchar* uniformName, GLuint programID, float set);
+		void setUniform(const GLchar* uniformName, GLuint programID, glm::vec3 set);
+		void setUniform(const GLchar* uniformName, GLuint programID, glm::vec4 set);
+		void setUniform(const GLchar* uniformName, GLuint programID, glm::mat4 set);
+	};
+
+	struct ShaderException : _exception {
```
Comment (iarba) [inline src/graphics/shader.h]: _exception yields exception
changing to std::exception fixed it for me

File: src/graphics/constants.h [hunk 5]
```diff
@@ -7,6 +7,7 @@ using uint = unsigned int;
 using ushort = unsigned short;
 using uchar = unsigned char;
 using llong = long long;
+const double pi = acos(-1);
```
Comment (iarba) [inline src/graphics/constants.h]: https://www.boost.org/doc/libs/1_37_0/libs/math/doc/sf_and_dist/html/math_toolkit/toolkit/internals1/constants.html

File: src/graphics/constants.h [hunk 6]
```diff
@@ -16,4 +17,17 @@ namespace graphics {
...
+
+	enum cameraMovements {
+		forward, backwards, up, down, left, right
+	};	
+
+	enum movementKeys {
+		forward = GLFW_KEY_W,
+		backwards = GLFW_KEY_S,
+		left = GLFW_KEY_A,
+		right = GLFW_KEY_D
```
Comment (iarba) [inline src/graphics/constants.h]: this doesn't compile - 
```
In file included from /home/iarba/Pof-MMORPG/src/graphics/graphics_files.h:3:0,
                 from /home/iarba/Pof-MMORPG/src/graphics/camera.h:3,
                 from /home/iarba/Pof-MMORPG/src/graphics/camera.cpp:1:
/home/iarba/Pof-MMORPG/src/include/glfw3.h:406:37: error: redeclaration of ‘forward’
 #define GLFW_KEY_W                  87
                                     ^
/home/iarba/Pof-MMORPG/src/graphics/constants.h:28:13: note: in expansion of macro ‘GLFW_KEY_W’
   forward = GLFW_KEY_W,
             ^
In file included from /home/iarba/Pof-MMORPG/src/graphics/variables.h:5:0,
                 from /home/iarba/Pof-MMORPG/src/graphics/graphics_files.h:4,
                 from /home/iarba/Pof-MMORPG/src/graphics/camera.h:3,
                 from /home/iarba/Pof-MMORPG/src/graphics/camera.cpp:1:
/home/iarba/Pof-MMORPG/src/graphics/constants.h:24:3: note: previous declaration ‘graphics::cameraMovements forward’
   forward, backwards, up, down, left, right
   ^
In file included from /home/iarba/Pof-MMORPG/src/graphics/graphics_files.h:3:0,
                 from /home/iarba/Pof-MMORPG/src/graphics/camera.h:3,
                 from /home/iarba/Pof-MMORPG/src/graphics/camera.cpp:1:
```

**Comment by iarba** [comment]
```
iarba@ubi16:~/Pof-MMORPG/build$ ./bin/gfx 
Compiling vertex shader: ../src/graphics/shaders/model.vertexshader
Compiling fragment shader: ../src/graphics/shaders/model.fragmentshader
Linking program
Cleaning up...
Segmentation fault (core dumped)
iarba@ubi16:~/Pof-MMORPG/build$ ./bin/gfx 
Compiling vertex shader: ../src/graphics/shaders/model.vertexshader
Compiling fragment shader: ../src/graphics/shaders/model.fragmentshader
Linking program
Cleaning up...
iarba@ubi16:~/Pof-MMORPG/build$ ./bin/gfx 
Compiling vertex shader: ../src/graphics/shaders/model.vertexshader
Compiling fragment shader: ../src/graphics/shaders/model.fragmentshader
Linking program
Cleaning up...
Segmentation fault (core dumped)
iarba@ubi16:~/Pof-MMORPG/build$ ./bin/gfx 
Compiling vertex shader: ../src/graphics/shaders/model.vertexshader
Compiling fragment shader: ../src/graphics/shaders/model.fragmentshader
Linking program
Cleaning up...
Segmentation fault (core dumped)
iarba@ubi16:~/Pof-MMORPG/build$ ./bin/gfx 
Compiling vertex shader: ../src/graphics/shaders/model.vertexshader
Compiling fragment shader: ../src/graphics/shaders/model.fragmentshader
Linking program
Cleaning up...
iarba@ubi16:~/Pof-MMORPG/build$ ./bin/gfx 
Compiling vertex shader: ../src/graphics/shaders/model.vertexshader
Compiling fragment shader: ../src/graphics/shaders/model.fragmentshader
Linking program
Cleaning up...
Segmentation fault (core dumped)
```
seems like you have a race condition somewhere

**Pull request merged**